From: Rob Evers <revers@redhat.com>
Date: Wed, 01 Dec 2010 16:44:33 -0000
Subject: [scsi] lpfc: fix panic in lpfc_scsi_cmd_iocb_cmpl
Message-id: <20101201164433.8352.65450.sendpatchset@localhost.localdomain>
Patchwork-id: 29769
O-Subject: [PATCH RHEL5.5z] lpfc: fix panic in lpfc_scsi_cmd_iocb_cmpl
Bugzilla: 658379
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=658379

Provided by Emulex.
Tested by Netapp and confirmed to resolved a panic in lpfc_scsi_cmd_iocb_cmpl

Previously accepted into rhel5.6 as part of Emulex lpfc update:

http://post-office.corp.redhat.com/archives/rhkernel-list/2010-August/msg01276.html

This needs to be ack'd tomorrow to make the next z-stream build and is blocking
qualification of rhel5.5z by netapp.
diff -urpN a/drivers/scsi/lpfc/lpfc_scsi.c b/drivers/scsi/lpfc/lpfc_scsi.c
--- a/drivers/scsi/lpfc/lpfc_scsi.c	2010-11-12 15:30:46.062617000 -0500
+++ b/drivers/scsi/lpfc/lpfc_scsi.c	2010-11-12 15:33:52.497798000 -0500
@@ -1486,15 +1486,21 @@ lpfc_scsi_cmd_iocb_cmpl(struct lpfc_hba 
 	struct lpfc_vport      *vport = pIocbIn->vport;
 	struct lpfc_rport_data *rdata = lpfc_cmd->rdata;
 	struct lpfc_nodelist *pnode = rdata->pnode;
-	struct scsi_cmnd *cmd = lpfc_cmd->pCmd;
+	struct scsi_cmnd *cmd;
 	int result;
 	struct scsi_device *tmp_sdev;
 	int depth;
 	unsigned long flags;
 	struct lpfc_fast_path_event *fast_path_evt;
-	struct Scsi_Host *shost = cmd->device->host;
+	struct Scsi_Host *shost;
 	uint32_t queue_depth, scsi_id;
 
+	/* Sanity check on return of outstanding command */
+	if (!(lpfc_cmd->pCmd))
+		return;
+	cmd = lpfc_cmd->pCmd;
+	shost = cmd->device->host;
+
 	lpfc_cmd->result = pIocbOut->iocb.un.ulpWord[4];
 	lpfc_cmd->status = pIocbOut->iocb.ulpStatus;
 	if (pnode && NLP_CHK_NODE_ACT(pnode))
